src/dune_rules/coq/coq_lib.ml

```diff
@@ -313,47 +475,90 @@ module DB = struct
+      | Stdlib lib -> Stdlib lib
+      | User_contrib lib -> User_contrib lib
+      | Not_found -> (
         match coq_db.parent with
-        | None -> None
+        | None -> Not_found
         | Some parent -> find parent name)
 
-    let find coq_db ~coq_lang_version name =
+    module Resolve_final_result = struct
+      (** Next we find corresponding Coq libraries for the various cases *)
+      type t =
+        | Found_lib of lib
+        | Found_stanza of Lib.DB.t * Path.Build.t * Coq_stanza.Theory.t
+        | Hidden
+        | Hidden_user_contrib
+        | Not_found
+    end
+
+    let find coq_db ~coq_lang_version name : Resolve_final_result.t =
       match find coq_db name with
-      | None -> `Not_found
-      | Some (mldb, dir, stanza) when coq_lang_version >= (0, 4) ->
-        `Found (mldb, dir, stanza)
-      | Some (mldb, dir, stanza) -> (
+      | Not_found -> Not_found
+      (* Composing with stdlib should be accepted *)
+      | Stdlib lib -> Found_lib lib
+      (* Composing with installed theories should come past 0.8 *)
+      | User_contrib lib when coq_lang_version >= (0, 8) -> Found_lib lib
+      | User_contrib _ -> Hidden_user_contrib
+      (* Composing with theories in the same project should come past 0.4 *)
+      | Theory (mldb, dir, stanza) when coq_lang_version >= (0, 4) ->
+        Found_stanza (mldb, dir, stanza)
+      | Theory (mldb, dir, stanza) -> (
         match coq_db.resolve name with
-        | `Not_found -> `Hidden
-        | `Theory _ | `Redirect _ -> `Found (mldb, dir, stanza))
-
-    let resolve_boot coq_db ~coq_lang_version =
-      let boot_id = boot_library_id coq_db in
-      resolve_boot ~coq_lang_version ~coq_db boot_id
+        | Not_found -> Hidden
+        | Theory _ | Redirect _ | Stdlib _ | User_contrib _ ->
+          Found_stanza (mldb, dir, stanza))
 
+    (** Our final final resolve is used externally, and should return the
+        library data found from the previous iterations. *)
     let resolve coq_db ~coq_lang_version (loc, name) =
       match find coq_db ~coq_lang_version name with
-      | `Not_found -> Error.theory_not_found ~loc name
-      | `Hidden -> Error.hidden_without_composition ~loc name
-      | `Found (db, dir, stanza) ->
+      | Not_found -> Error.theory_not_found ~loc name
+      | Hidden -> Error.hidden_without_composition ~loc name
+      | Found_stanza (db, dir, stanza) ->
         let open Memo.O in
         let+ theory = create_from_stanza coq_db db dir stanza in
         let open Resolve.O in
-        let* (_ : (Loc.t * Lib.t) list) = theory.libraries in
-        let+ (_ : (Loc.t * lib) list) = theory.theories in
+        let* (_ : (Loc.t * Lib.t) list) =
+          match theory with
+          | Dune t -> t.libraries
+          | Legacy _ -> Resolve.return []
+        in
+        let+ (_ : (Loc.t * lib) list) =
+          match theory with
+          | Dune t -> t.theories
+          | Legacy _ -> Resolve.return []
+        in
```

Comment:
Please move this to a function with a name that can be understood (and a comment)